components/director/internal/domain/runtime/service.go

```diff
@@ -525,39 +490,97 @@ func (s *service) ensureRuntimeExists(ctx context.Context, tnt string, runtimeID
+			return errors.Wrapf(err, "while assigning formation %q from runtime with ID %q", scenario, id)
+		}
 	}
 
-	scenarios, err := s.scenarioAssignmentEngine.MergeScenariosFromInputLabelsAndAssignments(ctx, newRuntimeLabels, runtimeID)
+	return nil
+}
+
+func (s *service) unassignRuntimeScenarios(ctx context.Context, rtmTenant, runtimeID string) error {
+	currentRuntimeLabels, err := s.getCurrentLabelsForRuntime(ctx, rtmTenant, runtimeID)
 	if err != nil {
-		return errors.Wrap(err, "while merging scenarios from input and assignments")
+		return err
 	}
 
-	if len(scenarios) == 0 {
-		err := s.labelRepo.Delete(ctx, rtmTenant, model.RuntimeLabelableObject, runtimeID, model.ScenariosKey)
+	if scenarios, areScenariosInLabels := currentRuntimeLabels[model.ScenariosKey]; areScenariosInLabels {
+		scenariosStr, err := label.ValueToStringsSlice(scenarios)
 		if err != nil {
-			return errors.Wrapf(err, "while deleting scenarios label from runtime with id [%s]", runtimeID)
+			return errors.Wrapf(err, "while converting scenarios: %+v to slice of strings", scenarios)
+		}
+
+		for _, scenario := range scenariosStr {
+			if _, err = s.formationService.UnassignFormation(ctx, rtmTenant, runtimeID, graphql.FormationObjectTypeRuntime, model.Formation{Name: scenario}); err != nil {
+				return errors.Wrapf(err, "while unassigning formation %q from runtime with ID %q", scenario, runtimeID)
+			}
 		}
-		return nil
 	}
 
-	scenariosLabelInput := &model.LabelInput{
-		Key:        model.ScenariosKey,
-		Value:      scenarios,
-		ObjectID:   runtimeID,
-		ObjectType: model.RuntimeLabelableObject,
+	return nil
+}
+
+func (s *service) updateScenariosLabel(ctx context.Context, rtmTenant, rtmId string, inputLabels map[string]interface{}) error {
+	mergedScenarios, err := s.formationService.MergeScenariosFromInputLabelsAndAssignments(ctx, inputLabels, rtmId)
+	if err != nil {
+		return errors.Wrap(err, "while merging scenarios from input and assignments")
 	}
 
-	err = s.labelUpsertService.UpsertLabel(ctx, rtmTenant, scenariosLabelInput)
+	mergedScenariosSlice, err := label.ValueToStringsSlice(mergedScenarios)
 	if err != nil {
-		return errors.Wrapf(err, "while creating scenarios label for Runtime with id [%s]", runtimeID)
+		return errors.Wrapf(err, "while converting merged scenarios: %+v to slice of strings", mergedScenarios)
+	}
+
+	currentRuntimeLabels, err := s.getCurrentLabelsForRuntime(ctx, rtmTenant, rtmId)
+	if err != nil {
+		return err
+	}
+
+	currentScenarios, areCurrentScenariosInLabels := currentRuntimeLabels[model.ScenariosKey]
+	if !areCurrentScenariosInLabels {
+		currentScenarios = []interface{}{}
+	}
+
+	currentScenariosSlice, err := label.ValueToStringsSlice(currentScenarios)
+	if err != nil {
+		return errors.Wrapf(err, "while converting current runtime scenarios: %+v to slice of strings", currentScenarios)
+	}
+
+	// Calculate the diff between current scenarios and merged scenarios.
```

Comment:
Maybe it will be easier to follow the method if you create two maps:
currentScenarios map[string]struct{} and mergedScenarios map[string]struct{}
then you can range currentScenarios and call unassignFormation for each scenario that is not found in mergedScenarios. After that you can range mergedScenarios and call assignFormation for each scenario that is not foun in currentScenarios.
This will reduce the complexity and nesting a lot and will be more straightforward approach.

Reply:
At first, this was my approach, but I came up with this second approach with the maps which I think is a little bit smarter, but if the code is not readable I will revert it.